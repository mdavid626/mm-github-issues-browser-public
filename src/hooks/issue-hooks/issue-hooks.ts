import { useQuery, gql } from '@apollo/client';
import { ApolloError } from '@apollo/client/errors';
import { IssueQueryResult, IssuesQueryResult } from '../../types/issue';

const issuesQuery = gql`
  query {
    repository(owner: "facebook", name: "react") {
      issues(last: 20) {
        edges {
          node {
            id
            createdAt
            title
            state
            number
            url
            author {
              login
            }
          }
        }
      }
    }
  }
`;

export const useIssues = (): [
  IssuesQueryResult | undefined,
  boolean,
  ApolloError | undefined
] => {
  const { data, loading, error } = useQuery<IssuesQueryResult>(issuesQuery);
  return [data, loading, error];
};

const issueQuery = gql`
  {
    repository(owner: "facebook", name: "react") {
      issue(number: 26077) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        bodyText
        author {
          login
        }
      }
    }
  }
`;

export const useIssue = (): [
  IssueQueryResult | undefined,
  boolean,
  ApolloError | undefined
] => {
  const { data, loading, error } = useQuery<IssueQueryResult>(issueQuery);
  return [data, loading, error];
};
