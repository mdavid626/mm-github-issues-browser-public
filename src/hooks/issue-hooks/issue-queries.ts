import { gql } from '@apollo/client';

export const GetIssuesQuery = gql`
  query getIssues($cursor: String, $search: String!) {
    search(query: $search, after: $cursor, type: ISSUE, first: 10) {
      issueCount
      nodes {
        ... on Issue {
          id
          createdAt
          title
          state
          number
          url
          author {
            login
          }
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
`;

export const GetIssueQuery = gql`
  query getIssue($issueNumber: Int!) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        body
        author {
          login
        }
        comments(first: 3, orderBy: { field: UPDATED_AT, direction: DESC }) {
          nodes {
            id
            createdAt
            body
            author {
              login
            }
          }
          pageInfo {
            hasNextPage
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;

export const GetIssueCommentsQuery = gql`
  query getIssueComments($issueNumber: Int!, $commentsCursor: String) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        comments(
          first: 3
          after: $commentsCursor
          orderBy: { field: UPDATED_AT, direction: DESC }
        ) {
          nodes {
            id
            createdAt
            body
            author {
              login
            }
          }
          pageInfo {
            hasNextPage
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;
