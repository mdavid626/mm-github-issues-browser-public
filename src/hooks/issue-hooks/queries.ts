import { gql } from '@apollo/client';

export const GetIssuesQuery = gql`
  query getIssues($cursor: String, $states: [IssueState!]) {
    repository(owner: "facebook", name: "react") {
      id
      issues(
        first: 10
        after: $cursor
        states: $states
        orderBy: { field: CREATED_AT, direction: DESC }
      ) {
        nodes {
          id
          createdAt
          title
          state
          number
          url
          author {
            login
          }
        }
        pageInfo {
          hasNextPage
          startCursor
          endCursor
        }
        totalCount
      }
    }
  }
`;

export const GetIssueQuery = gql`
  query getIssue($issueNumber: Int!) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        bodyText
        author {
          login
        }
        comments(first: 3, orderBy: { field: UPDATED_AT, direction: DESC }) {
          nodes {
            id
            createdAt
            bodyText
            author {
              login
            }
          }
          pageInfo {
            hasPreviousPage
            hasNextPage
            startCursor
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;

export const GetCommentsQuery = gql`
  query getIssueComments($issueNumber: Int!, $commentsCursor: String) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        comments(
          first: 3
          after: $commentsCursor
          orderBy: { field: UPDATED_AT, direction: DESC }
        ) {
          nodes {
            id
            createdAt
            bodyText
            author {
              login
            }
          }
          pageInfo {
            hasPreviousPage
            hasNextPage
            startCursor
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;
