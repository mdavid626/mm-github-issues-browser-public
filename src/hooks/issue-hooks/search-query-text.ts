import { Filters } from '../../types/filters';

const createSearchQueryText = (filters: Filters) => {
  const escapedSearch = filters.search.replace(/[^a-z\d\s]/gi, '');
  const search = filters.search
    ? `(in:title ${escapedSearch} OR in:body ${escapedSearch})`
    : '';
  const state = filters.state ? `state:${filters.state}` : '';
  return ['repo:facebook/react', 'type:issue', state, search]
    .filter((item) => item)
    .join(' ');
};

export default createSearchQueryText;
