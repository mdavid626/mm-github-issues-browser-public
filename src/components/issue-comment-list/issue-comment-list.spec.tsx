import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { useFetchMoreIssueComment } from '../../hooks/issue-hooks/issue-hooks';
import {
  testIssueComments1,
  testEmptyIssueComments,
} from '../../test-data/issue-comments';
import IssueCommentItem from '../issue-comment-item/issue-comment-item';
import IssueCommentList from './issue-comment-list';

jest.mock('../../hooks/issue-hooks/issue-hooks');
jest.mock('../issue-comment-item/issue-comment-item');

describe('issue-comment-list', () => {
  beforeEach(() => {
    (IssueCommentItem as jest.Mock).mockReturnValue(
      <div>issue-comment-item</div>
    );
    (useFetchMoreIssueComment as jest.Mock).mockReturnValue([jest.fn(), false]);
  });
  afterEach(cleanup);
  afterEach(jest.resetAllMocks);

  it('should render', () => {
    const { asFragment } = render(
      <IssueCommentList issueNumber={1} comments={testIssueComments1} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(useFetchMoreIssueComment).toHaveBeenCalledWith(1);
    expect(IssueCommentItem).toHaveBeenCalledTimes(2);
    expect(IssueCommentItem).toHaveBeenCalledWith(
      { comment: testIssueComments1.nodes[0] },
      {}
    );
    expect(IssueCommentItem).toHaveBeenCalledWith(
      { comment: testIssueComments1.nodes[1] },
      {}
    );
  });

  it('should render no comments message when no comments', () => {
    render(
      <IssueCommentList issueNumber={1} comments={testEmptyIssueComments} />
    );
    expect(screen.getByText('No comments yet')).toBeVisible();
    expect(IssueCommentItem).not.toHaveBeenCalled();
  });

  it('should fetch more comments when button clicked', async () => {
    const fetchMoreIssueComment = jest.fn();
    (useFetchMoreIssueComment as jest.Mock).mockReturnValue([
      fetchMoreIssueComment,
      false,
    ]);
    render(<IssueCommentList issueNumber={1} comments={testIssueComments1} />);
    await userEvent.click(screen.getByText('load more comments'));
    expect(fetchMoreIssueComment).toHaveBeenCalledWith(
      testIssueComments1.pageInfo.endCursor
    );
  });

  it('should show loading spinner when loading more comments', () => {
    (useFetchMoreIssueComment as jest.Mock).mockReturnValue([jest.fn(), true]);
    render(<IssueCommentList issueNumber={1} comments={testIssueComments1} />);
    expect(screen.getByText('loading...')).toBeVisible();
    expect(screen.queryByText('load more comments')).toBe(null);
  });
});
