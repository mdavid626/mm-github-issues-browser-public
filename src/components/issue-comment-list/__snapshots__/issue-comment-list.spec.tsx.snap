// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issue-comment-list should render 1`] = `
<DocumentFragment>
  <div
    class="IssueCommentList"
  >
    <div
      class="IssueCommentList-title"
    >
      Comments (2)
    </div>
    <div
      class="IssueCommentList-comments"
    >
      <div>
        issue-comment-item
      </div>
      <div>
        issue-comment-item
      </div>
    </div>
    <div
      class="IssueCommentList-loadMore"
    >
      load more comments
    </div>
  </div>
</DocumentFragment>
`;
